File: algorithm/cocktail-shaker-sort.js
/**
 * Cocktail Shaker Sort Implementation in JavaScript
 * 
 * Cocktail shaker sort (bidirectional bubble sort) is a variation of bubble sort
 * that sorts in both directions on each pass through the list.
 * 
 * Time Complexity: O(n²) worst/average, O(n) best (already sorted)
 * Space Complexity: O(1) - sorts in place
 */

function cocktailShakerSort(arr) {
    const sorted = [...arr];
    let swapped = true;
    let start = 0;
    let end = sorted.length - 1;
    
    while (swapped) {
        swapped = false;
        
        // Forward pass
        for (let i = start; i < end; i++) {
            if (sorted[i] > sorted[i + 1]) {
                [sorted[i], sorted[i + 1]] = [sorted[i + 1], sorted[i]];
                swapped = true;
            }
        }
        
        if (!swapped) break;
        
        swapped = false;
        end--;
        
        // Backward pass
        for (let i = end - 1; i >= start; i--) {
            if (sorted[i] > sorted[i + 1]) {
                [sorted[i], sorted[i + 1]] = [sorted[i + 1], sorted[i]];
                swapped = true;
            }
        }
        
        start++;
    }
    
    return sorted;
}

// Example usage
if (require.main === module) {
    const arr = [64, 34, 25, 12, 22, 11, 90];
    console.log('Cocktail Shaker Sort:', cocktailShakerSort(arr));
}

module.exports = { cocktailShakerSort };

File: algorithm/cocktail-shaker-sort.py
"""
Cocktail Shaker Sort Implementation in Python

Cocktail shaker sort (bidirectional bubble sort) is a variation of bubble sort
that sorts in both directions on each pass through the list.

Time Complexity: O(n²) worst/average, O(n) best (already sorted)
Space Complexity: O(1) - sorts in place
"""


def cocktail_shaker_sort(arr):
    sorted_arr = arr.copy()
    swapped = True
    start = 0
    end = len(sorted_arr) - 1
    
    while swapped:
        swapped = False
        
        # Forward pass
        for i in range(start, end):
            if sorted_arr[i] > sorted_arr[i + 1]:
                sorted_arr[i], sorted_arr[i + 1] = sorted_arr[i + 1], sorted_arr[i]
                swapped = True
        
        if not swapped:
            break
        
        swapped = False
        end -= 1
        
        # Backward pass
        for i in range(end - 1, start - 1, -1):
            if sorted_arr[i] > sorted_arr[i + 1]:
                sorted_arr[i], sorted_arr[i + 1] = sorted_arr[i + 1], sorted_arr[i]
                swapped = True
        
        start += 1
    
    return sorted_arr


if __name__ == "__main__":
    arr = [64, 34, 25, 12, 22, 11, 90]
    print("Cocktail Shaker Sort:", cocktail_shaker_sort(arr))

